primitives/rpc/debug/src/lib.rs

```diff
@@ -30,11 +30,11 @@ sp_api::decl_runtime_apis! {
 			extrinsics: Vec<Block::Extrinsic>,
 			transaction: &Transaction,
 			trace_type: single::TraceType,
-		) -> Result<single::TransactionTrace, sp_runtime::DispatchError>;
+		) -> Result<(), sp_runtime::DispatchError>;
```

Comment:
I was surprised to see `DispatchError` here. I usually associate that with submitting transactions. What sorts of things could go wrong in this call?

Reply:
Yeah you're right, in reality we are replaying one or more extrinsics that were pulled from the original block's body in the first place, so I'd say is not possible for any of them to return a `DispatchError` when applying it. Please correct me if I'm wrong.

However, I have a question:

- A user sends an Ethereum raw transaction that is preliminarily valid and is included in a block.
- A re-org happens and the Ethereum transaction that was originally and preliminarily valid in the now orphaned block is being re-scheduled to be included in a canonical one.
- The transaction is added to the re-validation queue and becomes invalid for whatever reason i.e. `Stale`.
- This will result on a `TransactionValidityError`.

That extrinsic will not be included in the block's body at all right? If that's the case we are good because:

- If a user wants to trace it by hash, that transaction will still be mapped to the non-canonical block in the secondary rocksdb and it will still valid in that context produced a valid - but not canonical - trace.
- If a user wants to trace a block in which that transactions was supposed to exist (but it doesn't because it became invalid in the new canonical chain), the transaction will not be there at all.

Reply:
So in "normal" FRAME transactions, a tx can be included in a block and still cause a dispatch error. A classic example is calling sudo from the wrong account. But I think this tracing is only about EVM transactions, and I don't think those can cause DipatchErrors. (right?)

If it's true that EVM transactions can't cause dispatch errors, then I agree with everything you wrote. Definitely I agree with your thinking about the re-org scenario. so in that case I think this function shouldn't even return result at all. just return `()`. And somewhere in its impl we could `.expect("Transactions being traced are in valid blocks. Transactions being traced are EVM transactions. Valid blocks don't contain evm transactions that cause DispatchError. Therefore this transaction does not cause a DispatchError. qed");`.

Quick sanity check though. When an EVM transaction runs out of gas, does that lead to `DispatchError`?

Reply:
I'm fine merging this either way though.

Reply:
> But I think this tracing is only about EVM transactions

It replays all transactions, FRAME and EVM, but only traces EVM ones. However we don't really capture errors for FRAME ones, as we are purely interested in how those affected the state in which the EVM transactions are executed over. I.e. Alice sent to Bob 1GLRM and right after, in the same block, Bob did something with his balance in an EVM contract.

> Can EVM transactions cause DipatchErrors?

Yes, https://github.com/paritytech/frontier/blob/master/frame/evm/src/runner/mod.rs#L26

> When an EVM transaction runs out of gas, does that lead to DispatchError?

I don't think is the case for out of gas, it's a EVM-specific error. But the runner can return DispatchError prior to starting the actual EVM execution (i.e. not enough balance to pay fees or the transaction cost etc.. like [here](https://github.com/paritytech/frontier/blob/master/frame/evm/src/runner/stack.rs#L89). Or after, for fee refunds like [here](https://github.com/paritytech/frontier/blob/master/frame/evm/src/runner/stack.rs#L107)